report/sections/TestRunner/chapter.tex

```diff
@@ -1,33 +1,32 @@
+However, if multiple users were to post a submission at the same time, the content of the file could potentially be overwritten while the GHC interpreter is running the tests.
 To address this, we generate a new directory with a unique name for each submission.
 Then, the user submission is written to two  \texttt{.hs} files inside this directory - a \texttt{code.hs} file containing the exercise submission, and a \texttt{test.hs} file containing the corresponding tests.
 This way, we can pass the path to this isolated directory to the GHC interpreter process as an argument.
 Once the tests have been run, the directory and its contents are deleted again.
 The \texttt{stdout} of the process is then read to determine whether the tests succeeded or not.
 Finally, the interpreter output and the test success flag is packaged into a JSON object and sent back to the client.
 
-Sequentially starting a new \texttt{runhaskell} process for each request was identified as a major bottleneck during testing, which will be described in detail in chapter \ref{chap:Benchmarking}.
-In order to address this, we started by refactoring the code base to allow running several instances of the GHC interpreter process at the same time.
-However, despite this running faster for a small number of clients, this approach in itself is not sufficiently stable:
-If multiple clients were to post a code submission at the same time, the Test Runner will start a new process for each request.
+Sequentially starting a new \texttt{runhaskell} process for each request is a major bottleneck when multiple users submit solutions in quick succession.
+Processing each request sequentially results in slow response times, which may also lead to connection timeouts.
+In order to address this,
+This was solved by allowing multiple instances of the GCH interpreter process to run at the same time.
+However, despite this being faster for a small number of clients, this approach in itself is not sufficient:
+If multiple clients post a code submission at the same time, the Test Runner will start a new process for each request.
 Given a large enough number of requests, this would crash the Test Runner web server.
 Therefore, we needed a way to limit the number of running GHC interpreter processes while at the same time storing subsequent requests for future processing.
 
+To solve the aforementioned problem, we decided to implement a queue system.
+The idea is to enqueue requests for future processing if the number of currently running GHC interpreter processes reaches an upper limit.
+This limit is defined as an environmental constant in the program.
+However, this introduces a new problem.
+The server can no longer directly respond to a POST request with the Test Runner results because the request may be handled at a later point in time.
+This means that if the request were enqueued for long enough, it would eventually time out.
```

Comment:
```suggestion
However, this introduced a new problem.
The server could no longer respond directly to a POST request with the Test Runner results because the request may be handled at a later point in time.
This means that if the request were enqueued for long enough, it would eventually time out.
```
tense